src/Cardano/Fee.hs

```diff
@@ -254,29 +267,94 @@ reduceSingleChange (Fee fee, Coin chng)
 --
--- Pre-condition 1: The given outputs list shouldn't be empty
--- Pre-condition 2: None of the outputs should be null
+-- == Pre-condition
 --
--- It returns the a list of pairs (fee, output).
-divvyFee :: Fee -> [Coin] -> [(Fee, Coin)]
-divvyFee _ outs | (Coin 0) `elem` outs =
-    error "divvyFee: some outputs are null"
-divvyFee (Fee f0) outs = go f0 [] outs
+-- Every coin in the given list must be __non-zero__.
+--
+-- == Examples
+--
+-- >>> distributeFee (Fee 2) [(Coin 1), (Coin 1)]
+-- [(Fee 1, Coin 1), (Fee 1, Coin 1)]
+--
+-- >>> distributeFee (Fee 4) [(Coin 1), (Coin 1)]
+-- [(Fee 2, Coin 1), (Fee 2, Coin 1)]
+--
+-- >>> distributeFee (Fee 7) [(Coin 1), (Coin 2), (Coin 4)]
+-- [(Fee 1, Coin 1), (Fee 2, Coin 2), (Fee 4, Coin 4)]
+--
+-- >>> distributeFee (Fee 14) [(Coin 1), (Coin 2), (Coin 4)]
+-- [(Fee 2, Coin 1), (Fee 4, Coin 2), (Fee 8, Coin 4)]
+--
+distributeFee :: Fee -> NonEmpty Coin -> NonEmpty (Fee, Coin)
+distributeFee _ outs | Coin 0 `F.elem` outs =
+    error "distributeFee: one or more coins has a value of zero."
+distributeFee (Fee feeTotal) coins =
+    NE.zip feesRounded coins
   where
-    total = (sum . map getCoin) outs
-    go _ _ [] =
-        error "divvyFee: empty list"
-    go fOut xs [out] =
-        -- The last one pays the rounding issues
-        reverse ((Fee fOut, out):xs)
-    go f xs ((Coin out):q) =
-        let
-            r = fromIntegral out / fromIntegral total
-            fOut = ceiling @Double (r * fromIntegral f)
-        in
-            go (f - fOut) ((Fee fOut, Coin out):xs) q
+    -- A list of rounded fee portions, where each fee portion deviates from the
+    -- ideal unrounded portion by as small an amount as possible.
+    feesRounded :: NonEmpty Fee
+    feesRounded
+        -- 1. Start with the list of ideal unrounded fee portions for each coin:
+        = feesIdeal
+        -- 2. Attach an index to each fee portion, so that we can remember the
+        --    original order:
+        & NE.zip indices
+        -- 3. Sort the fees into descending order of their fractional parts:
+        & NE.sortBy (comparing (Down . fractionalPart . snd))
+        -- 4. Apply pre-computed adjustments to each fee portion:
+        --    * fees with the greatest fractional parts are rounded up;
+        --    * fees with the smallest fractional parts are rounded down.
+        & NE.zipWith (\adj (idx, fee) -> (idx, floor fee + adj)) feeAdjustments
+        -- 5. Restore the original order:
+        & NE.sortBy (comparing fst)
+        -- 6. Strip away the indices:
+        & fmap (Fee . fromIntegral . snd)
+      where
+        indices :: NonEmpty Int
+        indices = 0 :| [1 ..]
+
+    -- A list of fee portion adjustments, with one adjustment per coin.
+    --
+    -- Since the ideal fee portion for each coin is a rational value, we must
+    -- therefore round the rational value to produce a final integer result.
+    --
+    -- * A fee adjustment of '1' indicates a fee portion to be rounded /up/.
```

Comment:
At first look I thought maybe RoundUp, RoundDown would be better, but 1, 0 is also clear here

Reply:
> At first look I thought maybe RoundUp, RoundDown would be better, but 1, 0 is also clear here

I really like this suggestion. :+1: 

I've applied the suggestion here: https://github.com/input-output-hk/cardano-coin-selection/pull/25/commits/a8ce44c9ac76a5eb0132d40887556cb9f34726b5